### DISCUSSION TITLE
how to make a dep between custom_target() and preprocess()?

### DISCUSSION THREAD

**Discussion by stsp** [mesonbuild/meson#12711] [Q&A]
I have this code:
```
plt_inc = custom_target('plt.inc',
  output: 'plt.inc',
  input: tct,
  command: [PD, '3', '@INPUT@'],
  capture: true)
cpp = meson.get_compiler('c')
sfiles = cpp.preprocess(['plt.S'],
  output: '@BASENAME@.s',
  dependencies: declare_dependency(
    compile_args: ['-D__ASSEMBLER__']))
```

Both targets work properly separately,
when started by hands in a needed order.
But I need to make sure `plt_inc` is built
before running the preprocessor for `plt.S`.
No matter what I do, this ordering is not
achieved.
How can I do that?

**Comment by stsp** [comment]
For example if I do this:
```
sfiles = cpp.preprocess('plt.S', plt_inc,
  output: '@BASENAME@.s',
  dependencies: declare_dependency(
    compile_args: ['-D__ASSEMBLER__']))
```
Or this:
```
sfiles = cpp.preprocess('plt.S',
  output: '@BASENAME@.s',
  dependencies: declare_dependency(
    compile_args: ['-D__ASSEMBLER__'],
    sources: plt_inc))
```

Then in both cases I get this:
```
ERROR: Multiple producers for Ninja target "src/preprocessor_0.p/plt.s". Please rename your targets.
```

Any other attempt to feed `plt_inc`
to preprocess() lead to other errors,
and no success.

**Reply by bruchar1** [reply]
I already answered in #12715. This is current not possible. We would need to add a `depends` keyword to `compiler.preprocess`. It is probably not there simply because nobody had a need for it before...

Alternatively, you could manually invoke the preprocessor using a `custom_target`.

**Comment by stsp** [comment]
Fixed in https://github.com/mesonbuild/meson/pull/12743
